specs/Standards/MultiToken/README.md

```diff
@@ -0,0 +1,539 @@
+    ///   contracts
+    ///
+    /// Arguments:
+    /// * `sender_id`: the sender of `mt_transfer_call`
+    /// * `previous_owner_id`: the account that owned the tokens prior to it being
+    ///   transferred to this contract, which can differ from `sender_id` if using
+    ///   Approval Management extension
+    /// * `token_ids`: the `token_ids` argument given to `mt_transfer_call`
+    /// * `msg`: information necessary for this contract to know how to process the
+    ///   request. This may include method names and/or arguments.
+    ///
+    /// Returns true if tokens should be returned to `sender_id`
+    fn mt_on_transfer(
+        &mut self,
+        sender_id: AccountId,
+        token_ids: Vec<TokenId>,
+        amounts: Vec<U128>,
+        msg: String,
+    ) -> PromiseOrValue<Vec<U128>>;
+}
+```
+### Resolver Trait
+#### Notes
+- TokenId is of type String
+```
+/// Used when MultiTokens are transferred using `mt_transfer_call`. This is the method that's called after `mt_on_transfer`. This trait is implemented on the MultiToken contract.
+pub trait MultiTokenResolver {
+    /// Finalize an `mt_transfer_call` chain of cross-contract calls.
+    ///
+    /// The `mt_transfer_call` process:
+    ///
+    /// 1. Sender calls `mt_transfer_call` on MultiToken contract
+    /// 2. MultiToken contract transfers token from sender to receiver
+    /// 3. MultiToken contract calls `mt_on_transfer` on receiver contract
+    /// 4+. [receiver contract may make other cross-contract calls]
+    /// N. MultiToken contract resolves promise chain with `mt_resolve_transfer`, and may
+    ///    transfer token back to sender
+    ///
+    /// Requirements:
+    /// * Contract MUST forbid calls to this function by any account except self
+    /// * If promise chain failed, contract MUST revert token transfer
+    /// * If promise chain resolves with `true`, contract MUST return token to
+    ///   `sender_id`
+    ///
+    /// Arguments:
+    /// * `previous_owner_id`: the owner prior to the call to `mt_transfer_call`
+    /// * `receiver_id`: the `receiver_id` argument given to `mt_transfer_call`
+    /// * `token_ids`: the `token_ids` argument given to `mt_transfer_call`
+    /// * `approvals`: if using Approval Management, contract MUST provide
+    ///   set of original approved accounts in this argument, and restore these
+    ///   approved accounts in case of revert. In this case it may be multiple sets of approvals
+    ///
+    /// Returns true if tokens were successfully transferred to `receiver_id`.
+    fn mt_resolve_transfer(
+        &mut self,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        token_ids: Vec<TokenId>,
+        amounts: Vec<U128>,
+    ) -> Vec<U128>;
+}
+```
+### Storage Management Trait
+#### Notes
+This is semi necessary for ft token types to be able to refund users for storage of many different token types like gold/silver... this might be slightly out of scope
```

Comment:
IMO, If we're about to merge this to be an actual standard, we need to decide if it is or isn't out of scope first, otherwise people reading the standard will be confused by the ambiguity.

If this functionality is fully implemented in the NEP141 path, IMO it should be supported in this specification also.

Reply:
I think we voted on this a while back to evict approval management and if it were necessary implement it later. This is a mistake I left in the spec from that time period. I'll yank it out and if we decide on approval management add comments back.

Reply:
approval management is likely to stay, after the last meeting going to resolve this for now, as the approval management part of the spec is up for review.